const createError = require("http-errors");
const mongoose = require("mongoose");
const Message = require("../models/messageM");

module.exports = {
  getAllMessages: async (req, res, next) => {
    try {
      const results = await Message.find({}, { __v: 0 });
      res.send(results);
    } catch (error) {
      next(error);
    }
  },

  createNewMessage: async (req, res, next) => {
    try {
      const message = new Message(req.body);
      const result = await message.save();
      res.status(201).send(result);
    } catch (error) {
      if (error.name === "ValidationError") {
        next(createError(422, error.message));
      } else {
        next(error);
      }
    }
  },

  findMessageById: async (req, res, next) => {
    try {
      const message = await Message.findById(req.params.id);
      if (!message) {
        throw createError(404, "Message does not exist in the database.");
      }
      res.send(message);
    } catch (error) {
      if (error instanceof mongoose.CastError) {
        next(createError(400, "Invalid Message Id"));
      } else {
        next(error);
      }
    }
  },

  updateAMessage: async (req, res, next) => {
    try {
      const result = await Message.findByIdAndUpdate(req.params.id, req.body, {
        new: true,
      });
      if (!result) {
        throw createError(404, "Message does not exist");
      }
      res.send(result);
    } catch (error) {
      if (error instanceof mongoose.CastError) {
        next(createError(400, "Invalid Message Id"));
      } else {
        next(error);
      }
    }
  },

  deleteAMessage: async (req, res, next) => {
    try {
      const result = await Message.findByIdAndDelete(req.params.id);
      if (!result) {
        throw createError(404, "Message does not exist.");
      }
      res.send(result);
    } catch (error) {
      if (error instanceof mongoose.CastError) {
        next(createError(400, "Invalid Message Id"));
      } else {
        next(error);
      }
    }
  },
};
